src/tools/rbd_mirror/InstanceWatcher.cc

```diff
@@ -349,6 +445,266 @@ void InstanceWatcher<I>::notify_image_release(
+    }
+
+    std::swap(sync_ctx->on_start, on_start);
+    sync_ctx->instance_id = sync_ctx->req->instance_id;
+    sync_ctx->req = nullptr;
+  }
+
+  on_start->complete(r == -ECANCELED ? r : 0);
+
+  if (r == -ECANCELED) {
+    notify_finish_sync(sync_ctx->sync_id);
+  }
+}
+
+template <typename I>
+void InstanceWatcher<I>::handle_notify_complete_sync(C_SyncRequest *sync_ctx,
+                                                     int r) {
+  dout(20) << "sync_id=" << sync_ctx->sync_id << ", r=" << r << dendl;
+
+  Context *on_complete = nullptr;
+  {
+    Mutex::Locker locker(m_lock);
+
+    auto it = m_remote_throttler_syncs.find(sync_ctx->sync_id);
+    if (it == m_remote_throttler_syncs.end()) {
+      dout(20) << "sync_id=" << sync_ctx->sync_id
+               << ": canceled on leader acquire" << dendl;
+    } else {
+      assert(it->second == sync_ctx);
+      m_remote_throttler_syncs.erase(it);
+    }
+    std::swap(sync_ctx->on_complete, on_complete);
+  }
+
+  if (on_complete != nullptr) {
+    on_complete->complete(0);
+  }
+}
+
+template <typename I>
+void InstanceWatcher<I>::print_sync_status(Formatter *f, stringstream *ss) {
+  dout(20) << dendl;
+
+  Mutex::Locker locker(m_lock);
+  if (m_instance_sync_throttler != nullptr) {
+    m_instance_sync_throttler->print_status(f, ss);
+  }
+}
+
+template <typename I>
+void InstanceWatcher<I>::handle_acquire_leader() {
+  dout(20) << dendl;
+
+  Mutex::Locker locker(m_lock);
+
+  assert(m_instance_sync_throttler == nullptr);
+  m_instance_sync_throttler = InstanceSyncThrottler<I>::create();
+
+  for (auto &it : m_remote_throttler_syncs) {
+    auto sync_ctx = it.second;
+    dout(20) << "sync_id=" << sync_ctx->sync_id
+             << ": canceling request to remote throttler" << dendl;
+    if (sync_ctx->req == nullptr) {
+      // It is waiting for notify_finish_sync. Clear instance_id so
+      // when notify_finish_sync is eventually called we will skip
+      // sending notification.
+      sync_ctx->instance_id.clear();
+      continue;
+    }
+    if (sync_ctx->on_start != nullptr) {
+      handle_sync_start(m_instance_id, it.first, sync_ctx->on_start);
+      sync_ctx->on_start = nullptr;
+    }
+    sync_ctx->req->cancel();
+  }
+  m_remote_throttler_syncs.clear();
+
+  m_leader_instance_id = m_instance_id;
+  unsuspend_notify_requests();
+}
+
+template <typename I>
+void InstanceWatcher<I>::handle_release_leader() {
+  dout(20) << dendl;
+
+  Mutex::Locker locker(m_lock);
+
+  assert(m_instance_sync_throttler != nullptr);
+
+  m_leader_instance_id.clear();
+
+  for (auto &it : m_local_throttler_syncs) {
+    auto &instance_id = it.first.first;
+    auto &sync_id = it.first.second;
+    auto &on_start = it.second;
+    dout(20) << "instance_id=" << instance_id << ", sync_id=" << sync_id
+             << ": canceling request to local throttler" << dendl;
+    if (on_start != nullptr) {
+      if (instance_id == m_instance_id) {
+        send_sync_start(sync_id, on_start);
```

Comment:
Nit: can you just bubble up -ESTALE for the local instance as well to avoid the special case?